File: tools/get_vep_consequences/get_vep_consequences.R
require(reshape2)
require(ggplot2)
require(dplyr)
require(readr)
require(rjson)

# For each LD-linked variant with top GWAS SNPs, list the variant's predicted effect from VEP

# TODO: Do it in batches by locus ... it's way to slow right now!

get_vep_consequences = function(config_file, input_ld_file, output_file)
{
	# Load config file
	config = fromJSON(file=config_file)$get_vep_consequences
	config$input_ld_file = input_ld_file
	config$output_file = output_file

	ld = load_ld_file(config)


	# Figure out which columns of the CADD file we need, to get the VEP consequence field
	skip_lines = 1
	if ("skip_cadd_lines" %in% names(config))
	{
		skip_lines = config$skip_cadd_lines
	}

	# TODO: Make this part more generalized to other consequence files
	head = system(sprintf("zcat %s | tail -n +%d | head -n 1", config$cadd_file, skip_lines+1), intern=TRUE)
	head = unlist(strsplit(head, "\t"))

	vep_index = which(head == "Consequence")
	ref_index = which(head == "Ref")
	alt_index = which(head == "Alt")
	gene_index = which(head == "GeneID")
	feature_index = which(head == "FeatureID")	

	# Initialize a data frame to hold all relevant SNPs
	max_total_snps = 1000000
	all_loci = as.data.frame(list(chr=rep(0, max_total_snps), pos=rep(0, max_total_snps), locus=rep(0, max_total_snps), vep = rep("", max_total_snps), ref=rep("", max_total_snps), alt=rep("", max_total_snps), gene=rep("", max_total_snps), feature=rep("", max_total_snps), r2=rep("", max_total_snps)), stringsAsFactors=FALSE)

	loc_table_idx = 0
	#for (i in 1:dim(ld)[1])
	for (i in 1:10)
	{
		print(i)
		snp = ld[i,]

		out = system(sprintf("tabix %s %s:%s-%s", config$cadd_file, as.character(snp$chr), 
					    as.character(snp$pos), as.character(snp$pos)), intern=TRUE)
	
		for (o in 1:length(out))
		{
			loc_table_idx = loc_table_idx + 1
			row = sapply(out[o], function(x) {as.character(strsplit(x, "\t")[[1]])})
			vep = row[vep_index]
			ref = row[ref_index]
			alt = row[alt_index]
			gene = row[gene_index]
			feature = row[feature_index]

			all_loci$chr[loc_table_idx] = snp$chr
			all_loci$pos[loc_table_idx] = snp$pos
			all_loci$r2[loc_table_idx] = snp$r2
			all_loci$locus[loc_table_idx] = snp$locus
			all_loci$vep[loc_table_idx] = vep
			all_loci$ref[loc_table_idx] = ref
			all_loci$alt[loc_table_idx] = alt
			all_loci$gene[loc_table_idx] = gene
			all_loci$feature[loc_table_idx] = feature
		}
	}

	all_loci = all_loci[1:loc_table_idx,]
	print(head(all_loci))
	print(tail(all_loci))

	# Write output files
	# One row per variant
	write.table(all_loci, file = config$output_file, sep="\t", row.names=FALSE, col.names=TRUE, quote=FALSE)       
}

load_ld_file = function(config)
{
	ld = read.table(config$input_ld_file, header=TRUE)

	return(ld)
}

args = commandArgs(trailingOnly=TRUE)

config_file = args[1]
input_ld_file = args[2]
output_file = args[3]

get_vep_consequences(config_file, input_ld_file, output_file)

File: tools/get_vep_consequences/get_vep_consequences.py
import sys
import json
import gzip

def get_vep_consequences(config_file, input_ld_file, output_file):
    
    with open(config_file) as f:
        config = json.load(f)["get_vep_consequences"]

    vars_to_keep = {}

    with open(input_ld_file) as f:
        f.readline()
        for line in f:
            data = line.strip().split()
            chrom = data[0]
            pos = data[1]
            locus = data[2]
            ld = data[3]
            vars_to_keep[(chrom, pos)] = (locus, ld)

    i = 0
    with open(output_file, "w") as w:
        w.write("chrom\tpos\tlocus\tld\tref\talt\tconsequence\tgene\ttranscript\thgnc\n")
        with gzip.open(config["cadd_file"]) as f:
            f.readline()
            f.readline()
            for line in f:
                i += 1
                if i % 1000000 == 0:
                    print i
                data = line.strip().split()
                chrom = data[0]
                pos = data[1]
                ref = data[2]
                alt = data[3]
                consequence = data[7]
                gene = data[18]
                transcript = data[19]
                hgnc = data[20]
                if ((chrom, pos)) not in vars_to_keep:
                    continue

                locus = vars_to_keep[(chrom, pos)][0]
                ld = vars_to_keep[(chrom, pos)][1]
                w.write("\t".join([chrom, pos, locus, ld, ref, alt, consequence, gene, transcript, hgnc]) + "\n")

config_file = sys.argv[1]
input_ld_file = sys.argv[2]
output_file = sys.argv[3]

get_vep_consequences(config_file, input_ld_file, output_file)
